1단계 미션 제출합니다.

baejunghyun36 opened the pull request:
# 04/09 (화)
- GET /index.html 응답하기
- CSS 지원하기
- Query String 파싱
- Redirect

기능은 완료된 상태이며, 회원가입 기능은 시간상 완료하지 못한 상태로 PR 요청하였습니다.

# 04/11 (목)
- 회원가입 기능 추가
- 리팩토링 진행

페어의 리뷰와 제가 받은 리뷰를 통합하여 수정사항 반영하여 PR 요청하였습니다. 
페어와의 공통 작업이라 1차 코드 리뷰건에 대한 수정사항은 페어의 리팩토링 코드를 Pull 받아 적용되어있고, 금일 페어와 수정한 코드를 분석할 예정입니다.

htrucci commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,5 @@
+package controller;
+
+public interface Controller {
+
```
2단계 요건중 아래와 같이 추상화하는 부분이 있는데 빠진 것 같네요~!
void service(HttpRequest request, HttpResponse response);

dongjoon1251 replied:
음 말씀해주신 부분은 2단계 요건은 아니고 힌트라서 맞추진 않았습니다.
Controller 인터페이스를 구현하고(사실 구현한다기 보단 @Controller 느낌으로 붙였는데 어노테이션으로는 구현하지 않았습니다 ㅎㅎ;)
HttpRequest, HttpResponse 파라미터를 받거나
HttpRequest request, HttpResponse response, Model model을 받는 메소드가 있으면
request mapping이 가능하도록 설계했습니다.

htrucci commented on src/main/java/webserver/request/HttpRequest.java:
```diff
@@ -0,0 +1,79 @@
+        return protocol;
+    }
+
+    public String getParam(String key) {
```
가능하다면 표준?으로 사용되는 getParameter로 명확히 명명하는 것은 어떨까요?

dongjoon1251 replied:
관련해서 Param으로 되어있는 부분 혹은 params 변수등은 모두 Parameter, parameters로 수정했습니다.

htrucci commented on src/main/java/controller/LoginController.java:
```diff
@@ -0,0 +1,37 @@
+                getRedirectPath(loginSuccess));
+    }
+
+    private String getCookieValue(boolean loginSuccess) {
```
Boolean.toString 으로 바로 리턴해도 되지 않을까 싶은데요.
getCookieValue 명칭이 조금 모호하다는 생각이 듭니다. 
좀 더 명확한 명칭을 쓰는게 좋을 것 같아요~!

dongjoon1251 replied:
https://developer.mozilla.org/ko/docs/Web/HTTP/Cookies
명칭이 어떤지 찾아보니 cookie-value라고 확인되었습니다.
옆에 LOGIN_COOKIE_KEY라고 한 부분은 해당 문서를 참고해서 LOGIN_COOKIE_NAME으로 수정했습니다.

htrucci replied:
아 그 cookieValue를 셋팅하는 부분이긴하나 login쿠키와 밀접하고 실제 함수의 역할을 고려하여 getLoginCookieValue 등으로 바꿔야하지 않나 생각이 들어 말씀드린 부분이었습니다~!

htrucci commented on src/main/java/controller/UserController.java:
```diff
@@ -0,0 +1,40 @@
+import webserver.request.HttpRequest;
+import webserver.response.HttpResponse;
+
+import static controller.LoginController.LOGIN_COOKIE_KEY;
```
개인적으로는 상수를 import하여 쓸때 static import하지 않고 
LoginController.LOGIN_COOKIE_KEY 와 같이 prefix를 붙여 사용하는데요.
Prefix로 인해 길어지긴 하지만 어디서 온 상수인지 가독이 쉬워서 유지보수하기 더 좋은 요소인 것 같아요.

LOGIN_COOKIE_KEY가 여러군데서 쓰일 가능성이 있기 때문에 쿠키키 상수를 가지는 별도의 constants클래스로 분리하고 
예로 Constants.LOGIN_COOKIE_KEY 와 같이 쓰는게 좋지않을지 개인적인 의견드립니다~

dongjoon1251 replied:
음 여러군데에서 쓰이지만 사용되는 모든 상수를 한곳에 모으면 constant 클래스는 더 많은 변경에 노출될 위험이 커보입니다. 
말씀해주신대로 LoginController. LOGIN_COOKIE_KEY 로 바꾸려다보니 사용되는 부분이 UserController밖에 없었습니다.
그래서 같은 상수를 사용해야하는 로직이면 합치는게 낫다고 생각이 들어서 UserController로  합치고, LoginController는 삭제했습니다.

htrucci commented on src/main/java/controller/LoginController.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    private String getRedirectPath(boolean loginSuccess) {
+        if (loginSuccess) {
```
삼항 연산자를 쓰면 더 깔끔할 것 같네요~!
return loginSuccess ? "/index.html" : "/user/login_failed.html";

htrucci commented on src/main/java/controller/UserController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public void addUser(HttpRequest request, HttpResponse response) {
+        userService.addUser(new User(request.getParam("userId"), request.getParam("password"), request.getParam("name"), request.getParam("email")));
```
User객체 생성하는 부분은 분리하는게 보기 편한 부분도 있을 것 같아요~!
User user = new User(request.getParam("userId"), request.getParam("password"), request.getParam("name"), request.getParam("email"))

dongjoon1251 replied:
method로 분리했습니다.

htrucci commented on src/main/java/controller/UserController.java:
```diff
@@ -0,0 +1,40 @@
+        return null;
+    }
+
+    private boolean isLogined(HttpRequest request) {
```
컨트롤러가 늘어난다고 하면 isLogined가 중복으로 발생할 것이어서,,
HttpRequest 로 옮겨저서 request.isLogined() 하면 되지않을까 생각이 들면서 
HttpRequest가 담당하는게 맞나 헷갈리는 부분이 있는데요. 혹시 의견이 있다면 피드백 부탁드립니다~!

dongjoon1251 replied:
'HTTP 요청/응답 처리 기능은 애플리케이션 개발자가 신경쓰지 않아도 재사용이 가능한 구조가 되도록 한다.'

위와 같은 요구사항이 있어서 애플리케이션 관련 로직은 controller, service 등에 있는게 더 낫다고 생각됩니다.
AuthenticationManager 같은 별도의 서비스 빈 클래스가 isLogined 역할을 담당하는게 더 맞지만 지금은 사용되는 곳이 한곳이라 분리하진 않았습니다.

htrucci commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -0,0 +1,70 @@
+
+    public void setRedirect(HttpRequest request, String path) {
+        this.header.setStatus(HttpStatus.FOUND);
+        this.header.setLocation("http://" + request.getHeader().getHost() + path);
```
찾아보니 풀path는 필요없을 것 같은데, 이유가 있을지 궁금합니다~

[https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Location]

dongjoon1251 replied:
1단계 미션 설명에 있는 https://en.wikipedia.org/wiki/HTTP_302 를 보고 다 붙였습니다. 
빼는게 더 깔끔하네요 👍

htrucci commented on src/main/java/controller/UserController.java:
```diff
@@ -0,0 +1,40 @@
+        }
+        response.setRedirect(request, "/user/login.html");
+        return null;
+    }
```
/user/list.html 호출시 showUsers를 타지 않는 것 같은데요. 
확인 부탁드립니다~
(templates) 는 전역적 템플릿 설정 경로이니 상수화하는 것이 어떨까요?

dongjoon1251 replied:
상수화하였습니다. 
비로그인시 진입되는 이슈는 이전에 로그인 후에 로컬 브라우저의 쿠키가 업데이트 된채로 남아있어서 그런 것 같습니다. 이부분은 3단계 미션에서 세션을 통해 개선될 수 있어 보입니다.

htrucci replied:
아 이건 제가 다시 확인해보니 /user/list 로 구현하고 /user/list.html 로 들어가란 요건은 없더라구요 ㅠ
user/list.html은 접근이 안되게 빼야할 것 같습니다.

htrucci commented on src/main/java/exception/ExceptionHandler.java:
```diff
@@ -0,0 +1,23 @@
+            logger.error(exception.getMessage());
+        }
+    }
+}
```
👍 👍 👍 👍 👍 
궁금해서 테스트를 해봤는데, 제가 방법이 틀렸을수도 있는데,, 
HttpResponse.error() 전에 소켓이 닫혀서 500으로 응답을 못하네요.

dongjoon1251 replied:
output stream이 return되고 나서는 닫히는 것 같네요.. exception handling하는 부분을 따로 메소드로 빼고 거기서 dispatch error를 처리하도록 수정했습니다.

htrucci commented on src/main/java/model/Model.java:
```diff
@@ -0,0 +1,28 @@
+    public Map<String, Object> getData() {
+        return data;
+    }
+}
```
👍 👍 👍 👍 👍

htrucci commented on src/main/java/utils/io/IOUtils.java:
```diff
@@ -0,0 +1,17 @@
+            br.read(body, 0, contentLength);
+        } catch (IOException exception) {
+            exception.printStackTrace();
+            throw new FileReadException();
```
예외처리를 추가해주셨네요!
테스트 코드를 작성하셔도 좋을 듯 합니다! 👍

htrucci commented on src/main/java/utils/template/TemplateUtils.java:
```diff
@@ -1,18 +1,15 @@
 import com.github.jknack.handlebars.io.TemplateLoader;
 
 public class TemplateUtils {
-    private static final String PREFIX = "/templates";
```
전역적인 templates경로가 바뀌면 viewPath를 다 바꿔줘야해서
PREFIX는 다시 와도 괜찮지 않을까 싶습니다.

dongjoon1251 replied:
다시 적용했습니다.

htrucci commented on src/main/java/webserver/Cookie.java:
```diff
@@ -1,36 +1,57 @@
 
-    public static Cookie login() {
-        return new Cookie("logined=true", "/");
+    public static Cookie fromRequest(String line) {
```
명확한 의미라서 좋긴하지만 정적 팩토리 메서드 컨벤션을 따라 그냥 from 혹은 of로 둬도 괜찮을 것 같아요~!

dongjoon1251 replied:
from으로 수정했습니다

htrucci commented on src/main/java/webserver/Cookie.java:
```diff
@@ -1,36 +1,57 @@
-        return new Cookie("logined=true", "/");
+    public static Cookie fromRequest(String line) {
+        String[] tokens = line.split("=");
+        return new Cookie(tokens[0].trim(), tokens[1].trim(), null);
```
로그인 쿠키설정시 Path를 /로 설정하는 부분이 리팩토링 과정에서 누락되었네요~!

dongjoon1251 replied:
요청으로부터 쿠키 객체를 만들때느 path정보가 없습니다. 응답에서 path 설정하는 부분은 다른 생성자를 사용했습니다.

htrucci commented on src/main/java/webserver/Dispatcher.java:
```diff
@@ -0,0 +1,92 @@
+        if (method.getParameterCount() == 2) {
+            invokeMethod(request, response, method);
+        }
+        if (method.getParameterCount() == 3) {
```
컨트롤러에 선언된 parameterCount로 체크하기에는 잠재적 위험이 있어보이네요.
parameterClassType을 읽어오거나 해서 매칭하여 model을 넘겨주는 것이 좀 더 안전할 것 같습니다~!

dongjoon1251 replied:
type도 체크하도록 수정했습니다

htrucci commented on src/main/java/webserver/FileController.java:
```diff
@@ -1,24 +0,0 @@
-import webserver.request.RequestPath;
-import webserver.response.Response;
-
-public class FileController {
```
구조가 분리되면서 컨트롤러를 없앴네요 👍

htrucci commented on src/main/java/webserver/WebServer.java:
```diff
@@ -1,14 +1,14 @@
 public class WebServer {
     private static final Logger logger = LoggerFactory.getLogger(WebServer.class);
-    private static final int DEFAULT_PORT = 8080;
+    private static final int DEFAULT_PORT = 8081;
```
8080포트가 요구사항에 있는데 포트가 달라졌네요~!

dongjoon1251 replied:
다시 8080으로 변경했습니다.

htrucci commented on src/main/java/controller/UserController.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public void addUser(HttpRequest request, HttpResponse response) {
+        userService.addUser(new User(request.getParam("userId"), request.getParam("password"), request.getParam("name"), request.getParam("email")));
```
빈값으로 생성이 가능하여, User 데이터 검증이 추가되면 좋을 것 같아요~!

dongjoon1251 replied:
1글자 이상인지 확인하는 로직을 User 생성자에 추가했습니다.

htrucci commented on src/main/java/webserver/request/HttpRequest.java:
```diff
@@ -0,0 +1,79 @@
+    private String path;
+    private Protocol protocol;
+    private RequestHeader header;
+    private Map<String, String> bodyParams;
```
queryString이 모두 파싱되어 parameterMap으로 들어오니 body는 빼도 되지 않을까 싶어요!

dongjoon1251 replied:
지금은 body에서 parameter만 받지만 
body를 통해서 처리할 수 있는 방식이 달라질 수 있을 것 같아서 필드로 분리해주는 게 낫다고 판단했습니다.

htrucci commented on src/main/java/webserver/request/RequestBuilder.java:
```diff
@@ -0,0 +1,56 @@
+    private static boolean isCookie(String key) {
+        return key.equals(COOKIE);
+    }
+}
```
HttpRequest를 만들때만 사용하는 클래스이다 보니 
내부 클래스로 포함시켜서 정말 Builder패턴의 Builder가 되었어도 괜찮았을 것 같아요~!
혹시 안티패턴이면 말씀해주세요~

dongjoon1251 replied:
Builder로 넣을까 하다가 HttpRequest가 커지기도 했고, 토크나지잉 로직도 있어서 분리시켰습니다. 안티패턴이라기보다는 클래스명을 Builder로 넣어서 그런것 같습니다. 해당 클래스 역할은 parsing에 더 가까워서 RequestParser로 이름을 변경했습니다.

htrucci commented on src/main/java/webserver/requestmapping/RequestMapping.java:
```diff
@@ -0,0 +1,67 @@
+        map = new HashMap<>();
+        map.put("/user/create", getMethodMap(userController,
+                Arrays.asList(HttpMethod.POST, HttpMethod.GET),
+                Arrays.asList("addUser", "addUser")));
```
addUser가 두개 들어가는건 필요한 부분인가요?

dongjoon1251 replied:
메소드별로 다른 로직을 태울 가능성도 있어서 일단 중복으로 넣어놨습니다.

htrucci commented on src/main/java/webserver/requestmapping/RequestMapping.java:
```diff
@@ -0,0 +1,67 @@
+        map.put("/user/list", getMethodMap(userController,
+                Collections.singletonList(HttpMethod.GET),
+                Collections.singletonList("showUsers")));
+    }
```
실제 RequsetMappingHandler와 비슷하게 잘 짜셨는데 메서드명 변경시 잠재적 오류가 있을 것 같네요 ㅠㅠ

힌트요건에 하나의 컨트롤러 인터페이스를 상속받아 3가지 컨트롤러로 분리하도록 구현하도록 되어있긴한데,,
힌트 내용과 같이 동일한 메서드를 사용하여 맵핑이 필요없다면 아래와 같이 선언 및 추가가 가능하여 구조적으로 더 좋을 것 같습니다.

USER_CREATE(Method.GET, "/user/create", CreateUserController.class)
USER_LOGIN(Method.POST, "/user/login", LoginController.class)
USER_LIST(Method.GET, "/user/list", ListUserController.class)

dongjoon1251 replied:
enum을 컴파일 하는 시점에 method가 정확히 있는지 체크하는 로직을 추가했습니다. 그리고 말씀해주신 것처럼 enum으로 바꾸고 나니 훨씬 깔끔해졌네요 :)

dongjoon1251 replied:
어노테이션 통해서 requestMapping에 추가되도록 하는 부분은 2번째 미션에 있는 것 같아서 추가하진 않았습니다 :)

htrucci commented on src/main/java/webserver/response/ResponseHeader.java:
```diff
@@ -1,124 +1,65 @@
-    private void setLocation(String location) {
-        this.location = location;
+    public void setContentLength(int contentLength) {
+        headers.add(String.format("Content-Length: %d\r\n", contentLength));
```
String.format을 활용한 부분이 좋은 것 같습니다~! 👍

htrucci commented on src/test/java/webserver/request/RequestBuilderTest.java:
```diff
@@ -0,0 +1,83 @@
+                "Host: localhost:8080 \n" +
+                "Connection: keep-alive \n" +
+                "Accept: */*\n";
+
```
step2 힌트내용과 같이 요청 내용을 파일로 만들어놓고 읽어서 쓰면 테스트 코드 활용성이 더 높아질 것 같아요~!

`InputStream in = new FileInputStream(new File(testDirectory + "Http_POST2.txt"));
  HttpRequest request = new HttpRequest(in);`

dongjoon1251 replied:
test 폴더에 resource 폴더를 추가해서 말씀해주신 방법으로 테스트 수정했습니다

htrucci reviewed (changes requested):
안녕하세요~
2단계 구현하느라 고생하셨습니다~! 역할분리 및 mvc구조화가 아주 잘 되었네요! 👍
몇가지 의견 확인해주시고 다시 리뷰 요청 부탁드립니다~!

htrucci commented on src/main/java/webserver/Cookie.java:
```diff
@@ -4,27 +4,27 @@
 public class Cookie {
 
-    private final String key;
+    private final String name;
```
아! name이 더 올바른 네임이네요! ㅎㅎ

baejunghyun36 committed:
feat : Uri 클래스 구현

baejunghyun36 committed:
feat : HttpVersion 클래스 구현

baejunghyun36 committed:
feat : Method 객체를 별도로 구성
- Uri의 변경사항도 함께 반영

baejunghyun36 committed:
chore : Uri에서 Method를 분리

baejunghyun36 committed:
feat : Uri 생성 로직 추가
- 생성 방식 변경
- 파싱 로직 추가

baejunghyun36 committed:
feat : HttpVersion 생성 로직 추가
- 생성 방식 변경
- 파싱 로직 추가

baejunghyun36 committed:
feat : RequestLine 생성 로직 추가
- 생성 방식 변경
- 파싱 로직 추가

baejunghyun36 committed:
feat : Uri 검증 로직 추가

baejunghyun36 committed:
feat : HttpHeaders 생성 로직 추가
- 생성 방식 변경
- 파싱 로직 추가

baejunghyun36 committed:
test : HttpHeaders 테스트 보완

baejunghyun36 committed:
feat : RequestLine 생성 방식 변경

baejunghyun36 committed:
fix : inputstream 대신 bufferedreader를 주입받는 방식으로 변경 1

baejunghyun36 committed:
fix : inputstream 대신 bufferedreader를 주입받는 방식으로 변경 2

baejunghyun36 committed:
feat : httprequest 클래스 생성

baejunghyun36 committed:
chore : 불필요한 주석 제거

baejunghyun36 committed:
chore : 테스트 클래스 접근지정자 제거

baejunghyun36 committed:
feat : requestbody 관련 로직 추가
- FormUrlEncoded 방식 파싱 로직 추가
- HttpHeaders의 get 메서드 원형 변경
- HttpRequest 구성 요소인 RequestBody 정의
- RequestBodyParsingStrategy의 메서드 분리

baejunghyun36 committed:
refactor : 패키지 구조 수정

baejunghyun36 committed:
feat : StatusLine 클래스 구현

baejunghyun36 committed:
chore : static import로 변경

baejunghyun36 committed:
fix : HttpResponse 구조 변경

baejunghyun36 committed:
docs : 기능 목록 작성

baejunghyun36 committed:
feat : 컨트롤러 등록 구조 작성

baejunghyun36 committed:
feat : 웹 서버 베이스 기능 구현

Commit:
chore : HttpMethod -> Method, 미완성 테스트 코드 완성

Commit:
refactor : WebApplication 구조 변경 및 비효율 구문 변경
- 별도 메서드 분리 및 else 제거
- 불필요한 테스트 제거
- 매 루프마다 UrlDecoding하지 않고 한 번에 Decoding 후 처리로 변경

Commit:
refactor : 매직넘버에 이름을 부여
- 컨트롤러 중복으로 서버 기동 불가 문제와 함께 처리

Commit:
chore : 예외 문구 대소문자 변경, header 값 직접 추출 불가 변경

Commit:
refactor : 계층 별 캡슐화
- HttpHeaders에서 접근하여 값을 빼내지 않도록 변경

Commit:
fix : 테스트 코드 통과

Commit:
fix : 쿼리 파라미터 유효성 검증 방식 변경

Commit:
refactor : Factory를 이용하는 방식으로만 파싱 전략 생성 가능하도록 변경

Commit:
refactor : Spring Framework 의존성 제거

Commit:
refactor : HttpResponse 초기 구성
- HttpVersion의 불필요한 퀄리파이어 삭제
- HttpHeaders의 생성 방식 추가
- 기존 HttpResponse를 Render로 전환
- HttpResponse가 StatusLine과 HttpHeaders를 갖도록 변경
- 임시 조치로 추가 변경 필요 (StatusCode 정의, Renderer 리팩토링 등)

Commit:
refactor : Enum 리패키징

Commit:
feat : StatusCode를 정의

Commit:
feat : ResponseBody를 정의하고 HttpResponse가 이를 갖도록 추가

Commit:
test : 테스트 코드 누락된 부분 커밋

Commit:
feat : HttpResponse가 Render 되도록 로직 구현

Commit:
refactor : RequestMapper가 갖고 있던 StaticServing을 별도 객체로 분리

Commit:
feat : 사용자 회원 가입 로직 구현 (GET / POST)

hyunssooo commented on src/main/java/webserver/common/HttpVersion.java:
```diff
@@ -0,0 +1,100 @@
+    private final String raw;
+
+    public HttpVersion(String version) {
+        (new HttpVersion.Parser(version)).parse();
```
이 구조가 조금 어색한 것 같은데요. Parser 객체가 HttpVersion의 필드들을 할당하고 있네요. 
Parser 에서 HttpVersion 필드들을 할당하는 이유가 따로 있나요? 
parse 메서드 내부에서 사용하고 있는 validateSchemeChunk() 은 또 HttpVersion 내부에 있네요.

hyunssooo commented on src/main/java/webserver/controller/Controller.java:
```diff
@@ -0,0 +1,9 @@
+import webserver.request.HttpRequest;
+import webserver.response.HttpResponse;
+
+@FunctionalInterface
+public interface Controller {
+    void doService(HttpRequest req, HttpResponse response);
+}
```
추상화 👍

hyunssooo commented on src/main/java/webserver/WebApplicationServer.java:
```diff
@@ -30,4 +31,11 @@ public static void main(String args[]) throws Exception {
+        });
+        RequestMapper.register(HttpMethod.POST.name(), "/user/create", (httpRequest, response) -> {
+        });
+    }
```
controller의 구현체는 따로 없나요? 

https://edu.nextstep.camp/s/MCLQmhAp/ls/z7W5Kvcv
요구 사항에 POST 방식으로 회원가입이 있는데요 이 부분은 구현되어 있나요?

hyunssooo commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -0,0 +1,57 @@
+import java.io.OutputStream;
+
+public final class HttpResponse {
+    private static final Logger logger = LoggerFactory.getLogger(RequestHandler.class);
```
HttpResponse 도 HttpRequest 와 마찬가지로 객체로 만들어 보는 것은 어떨까요?

hyunssooo commented on src/test/java/webserver/controller/RequestMapper.java:
```diff
@@ -0,0 +1,32 @@
+
+public final class RequestMapper {
+//    private static final Map<RequestLine>
+
```
이 클래스는 사용하지 않고 test에 존재하고 있네요. 제거해도 괜찮지 않을까요?

hyunssooo commented on src/main/java/webserver/controller/RequestMapper.java:
```diff
@@ -0,0 +1,100 @@
+    private static final String TEMPLATE_PATH = "./templates";
+    private static final String STATIC_PATH = "./static";
+
+    public static void register(String method, String path, Controller controller) {
```
사용하는 부분에서 Method.name()을 사용하고 내부에서 다시 Method를 사용하고 있네요. 이 부분은 그냥 Method를 받으면 안되나요? 의도하신 바가 있을까요?

hyunssooo commented on src/main/java/webserver/request/RequestBodyParsingStrategy.java:
```diff
@@ -0,0 +1,9 @@
+package webserver.request;
+
+import java.util.Map;
+
+public interface RequestBodyParsingStrategy {
+    void perform();
```
전략 추상화 👍

hyunssooo commented on src/test/java/webserver/common/HttpHeadersTest.java:
```diff
@@ -0,0 +1,29 @@
+        HttpHeaders headers = new HttpHeaders(br);
+        assertThat(headers.get("Content-Type").isEmpty()).isTrue();
+    }
+}
```
HttpHeaders 의 생성자가 Map 을 받는다면 좀 더 유연한 테스트가 가능하지 않을까요?

hyunssooo commented on src/test/java/webserver/common/HttpVersionTest.java:
```diff
@@ -0,0 +1,41 @@
+    void 불완전한_버전_표기에도_RuntimeException이_발생한다() {
+        String version = "HTTP/1.";
+        assertThatThrownBy(() -> new HttpVersion(version)).isInstanceOf(IllegalArgumentException.class);
+    }
```
꼼꼼한 테스트 좋습니다 :)

hyunssooo commented on src/main/java/webserver/common/HttpVersion.java:
```diff
@@ -0,0 +1,100 @@
+package webserver.common;
+
+import java.util.Objects;
+
+public final class HttpVersion {
```
자주 만들어지는 객체 같은데요. 미리 생성해두고 재 사용할 수 있지 않을까요?

hyunssooo reviewed (changes requested):
안녕하세요. 🙂
이번 미션 함께 진행할 정현수입니다.
몇 가지 코멘트 남겼어요. 확인 부탁드립니다!
함께 고민해 보고 싶은 부분은 코멘트나 DM 으로 남겨주세요!
늦어서 죄송합니다. 🙇‍♂️

Commit:
refactor : 페어의 1차 피드백을 반영
- HttpVersion이 매번 생성되지 않도록 static 객체 정의
- HttpVersion의 파싱을 Parser가 담당하도록 변경
- Method의 문자열이 아니라 객체를 활용
- Headers의 생성 방식을 추가 및 테스트

Commit:
docs: 기능 목록 업데이트

Commit:
refactor : Handlebars 템플릿을 지원하도록 변경

hyunssooo commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -0,0 +1,78 @@
+    public static final String DEFAULT_ENTRY = "/index.html";
+    private static final String EMPTY = "";
+    private static final String NOT_FOUND_MESSAGE = "NOT FOUND";
+
+    private final HttpRequest httpRequest;
```
HttpResponse.httpRequest 는 따로 사용하지 않는데요. 가지고 있을 필요 있나요?

baejunghyun36 replied:
넵 삭제하도록 하겠습니다!

hyunssooo commented on src/main/java/webserver/request/HttpRequest.java:
```diff
@@ -0,0 +1,97 @@
+    private final RequestLine requestLine;
+    private final HttpHeaders httpHeaders;
+    private final RequestBody requestBody;
+    private Object attribute;
+
```
사용하고 있을까요? 사용하지 않는다면 제거해도 괜찮을 것 같아요. 🙂

hyunssooo commented on src/main/java/webserver/request/HttpRequest.java:
```diff
@@ -0,0 +1,97 @@
+    private final RequestBody requestBody;
+    private Object attribute;
+
+    public HttpRequest() {
+        this.requestLine = null;
+        this.httpHeaders = null;
+        this.requestBody = null;
+    }
```
테스트를 위한 기본 생성자로 보이네요. 여러 생성자를 지원하면 사용하는 입장에서 선택지를 넓혀준다고 생각하지만 이 생성자는 큰 도움은 안될 것 같아요. HttpResponse 에서 httpRequest를 제거해도 괜찮지 않을까요?

baejunghyun36 replied:
삭제하겠습니다~

hyunssooo commented on src/main/java/webserver/request/HttpRequest.java:
```diff
@@ -0,0 +1,97 @@
+        this.requestBody = null;
+    }
+
+    public HttpRequest(InputStream in) throws IOException {
+        BufferedReader br = new BufferedReader(new InputStreamReader(in));
+        this.requestLine = new RequestLine(br);
+        this.httpHeaders = new HttpHeaders(br);
+        if (!isRequestBodyAcceptable()) {
+            this.requestBody = new RequestBody();
+            return;
+        }
+        this.requestBody = new RequestBody(RequestBodyParsingStrategyFactory.create(
+                contentType(),
+                IOUtils.readData(br, contentLength())));
+    }
```
위와 동일한 코멘트인데 여러 생성자를 두고 부 생성자는 주 생성자를 사용하게 하는 것은 어떨까요? 여러 생성자를 통해 사용하는 입장에서도 더 쉽게 사용할 수 있고 테스트도 더 쉽게 가능하지 않을까요? 
혹은 이 경우에는 정적 팩터리 메서드를 사용하는 것도 방법이겠네요! 
이펙티브 자바 아이템 1. 생성자 대신 정적 팩터리 메서드를 고려하라 를 참고해 봐도 좋을 것 같아요!

baejunghyun36 replied:
정적 팩터리 메서드를 활용하도록 수정하겠습니다~

hyunssooo commented on src/main/java/webserver/request/HttpRequest.java:
```diff
@@ -0,0 +1,97 @@
+    public Object attribute() {
+        return attribute;
+    }
+
+    public void setAttribute(Object attribute) {
+        this.attribute = attribute;
+    }
```
현재 사용하지 않는다면 과감히 제거하는 것은 어떨까요?

hyunssooo commented on src/test/java/webserver/common/HttpHeadersTest.java:
```diff
@@ -0,0 +1,42 @@
+class HttpHeadersTest {
+    @Test
+    void HttpHeaders를_Map을_이용하여_생성() {
+        Map<String, String> map = new HashMap<>() {{
+            put("Host", "localhost:8080");
+            put("Connection", "keep-alive");
+            put("Accept", "*/*");
+        }};
```
자바 9 이상에서는 `Map#of` 도 지원하고 있어요. :)

baejunghyun36 replied:
넵 활용해볼게요!

hyunssooo commented on src/main/java/webserver/controller/PostUserJoinController.java:
```diff
@@ -0,0 +1,23 @@
+import webserver.enums.MediaType;
+import webserver.request.HttpRequest;
+import webserver.response.HttpResponse;
+
+public class PostUserJoinController implements Controller {
+    @Override
+    public void doService(HttpRequest req, HttpResponse response) {
```
비즈니스 로직을 담당할 Controller 좋습니다 :)

hyunssooo reviewed (approved):
안녕하세요. 
전체적으로 깔끔하게 잘 구성해 주셨네요! 
몇 가지 추가로 코멘트 남겼습니다. 
다음 step 진행하시면서 확인 부탁드려요!
